system/jlib/jfile.cpp

```diff
@@ -2049,13 +2070,23 @@ void CFileIO::close()
 
 void CFileIO::flush()
 {
+    if (0 == (extraFlags & (IFEnocache | IFEsync)))
+        return;
+
     CriticalBlock procedure(cs);
+
+    if (extraFlags & IFEsync)
+    {
 #ifdef F_FULLFSYNC
-    if (fcntl(file, F_FULLFSYNC) != 0)
+        if (fcntl(file, F_FULLFSYNC) != 0)
 #else
-    if (fdatasync(file) != 0)
+        if (fdatasync(file) != 0)
 #endif
-        throw makeOsException(DISK_FULL_EXCEPTION_CODE, "CFileIO::flush");
+            throw makeOsException(DISK_FULL_EXCEPTION_CODE, "CFileIO::flush");
+    }
+    else
+        sync_file_region(file, 0, 0, false);
```

Comment:
similar question to previous one in close(). Is sync_file_region necessary for IFEnocache semantics, what would it mean if only posix_fadvise(fPOSIX_FADV_DONTNEED) was issued?

Reply:
You are then at the mercy of the kernel as to when the dirty blocks get committed and marked as clean.